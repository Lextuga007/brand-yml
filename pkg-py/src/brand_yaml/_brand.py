from __future__ import annotations

from typing import Any, List, Literal, Union

from pydantic import BaseModel, ConfigDict, HttpUrl, RootModel


class Brand(BaseModel):
    model_config = ConfigDict(extra="ignore")

    meta: BrandMeta = None
    logo: str | BrandLogo = None
    color: BrandColor = None
    typography: BrandTypography = None
    defaults: dict[str, Any] = None

class BrandStringLightDark(BaseModel):
    model_config = ConfigDict(extra="ignore", str_strip_whitespace=True)

    light: str | dict = None
    dark: str | dict = None


class BrandLogo(BaseModel):
    model_config = ConfigDict(extra="forbid")

    with_: dict[str, str | BrandStringLightDark] = None
    small: str | BrandStringLightDark = None
    medium: str | BrandStringLightDark = None
    large: str | BrandStringLightDark = None


type BrandColorValue = str


class BrandColor(BaseModel):
    model_config = ConfigDict(extra="forbid")

    with_: dict[str, BrandColorValue] = None
    foreground: BrandColorValue = None
    background: BrandColorValue = None
    primary: BrandColorValue = None
    secondary: BrandColorValue = None
    tertiary: BrandColorValue = None
    success: BrandColorValue = None
    info: BrandColorValue = None
    warning: BrandColorValue = None
    danger: BrandColorValue = None
    light: BrandColorValue = None
    dark: BrandColorValue = None
    emphasis: BrandColorValue = None
    link: BrandColorValue = None


type BrandNamedColorType = Literal[
    "foreground",
    "background",
    "primary",
    "secondary",
    "tertiary",
    "success",
    "info",
    "warning",
    "danger",
    "light",
    "dark",
    "emphasis",
    "link",
]
BrandNamedColorValues: BrandNamedColorType = (
    "foreground",
    "background",
    "primary",
    "secondary",
    "tertiary",
    "success",
    "info",
    "warning",
    "danger",
    "light",
    "dark",
    "emphasis",
    "link",
)


class BrandTypography(BaseModel):
    with_: BrandFont | list[BrandFont] = None
    base: BrandTypographyOptions = None
    headings: BrandTypographyOptionsNoSize = None
    monospace: BrandTypographyOptions = None
    emphasis: BrandTypographyEmphasis = None
    link: BrandTypographyLink = None


class BrandTypographyOptions(BaseModel):
    family: str = None
    size: str = None
    line_height: str = None
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = None
    style: BrandTypographyFontStyleType | list[BrandTypographyFontStyleType] = None
    color: str | BrandNamedColorType = None
    background_color: str | BrandNamedColorType = None


class BrandTypographyOptionsNoSize(BaseModel):
    family: str = None
    line_height: str = None
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = None
    style: BrandTypographyFontStyleType | list[BrandTypographyFontStyleType] = None
    color: str | BrandNamedColorType = None
    background_color: str | BrandNamedColorType = None


class BrandTypographyEmphasis(BaseModel):
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = None
    color: str | BrandNamedColorType = None
    background_color: str | BrandNamedColorType = None


class BrandTypographyLink(BaseModel):
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = None
    decoration: str = None
    color: str | BrandNamedColorType = None
    background_color: str | BrandNamedColorType = None


type BrandTypographyFontWeightType = Literal[
    100, 200, 300, 400, 500, 600, 700, 800, 900
]
BrandTypographyFontWeightValues: BrandTypographyFontWeightType = (
    100,
    200,
    300,
    400,
    500,
    600,
    700,
    800,
    900,
)


type BrandTypographyFontStyleType = Literal["normal", "italic"]
BrandTypographyFontStyleValues: BrandTypographyFontStyleType = ("normal", "italic")


class BrandTypographyFontGoogle(BaseModel):
    family: str
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = 400
    style: BrandTypographyFontStyleType | list[BrandTypographyFontStyleType] = "normal"
    display: Literal["auto", "block", "swap", "fallback", "optional"] = "auto"


class BrandFontFile(BaseModel):
    model_config = ConfigDict(extra="forbid")
    family: str
    files: str | list[str] = None


class BrandFontFamily(RootModel):
    root: str


class BrandFont(RootModel):
    root: List[Union[BrandTypographyFontGoogle, BrandFontFile, BrandFontFamily]]
