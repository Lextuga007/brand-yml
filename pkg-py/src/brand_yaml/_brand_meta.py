from __future__ import annotations

from pydantic import BaseModel, ConfigDict, HttpUrl, Field


class BrandMeta(BaseModel):
    """
    Brand metadata is stored in `meta`, providing place to describe the company
    or project, the brand guidelines, additional links, and more.
    """

    model_config = ConfigDict(extra="allow", str_strip_whitespace=True)

    name: str | BrandMetaName = Field(
        None, examples=["Very Big Corporation of America"]
    )
    link: HttpUrl | BrandLink = Field(
        None,
        examples=[
            "https://very-big-corp.com",
            '{"home": "https://very-big-corp.com"}',
        ],
    )


class BrandMetaName(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    full: str = Field(None, examples=["Very Big Corporation of America"])
    short: str = Field(None, examples=["VBC"])


class BrandLink(BaseModel):
    model_config = ConfigDict(extra="allow", str_strip_whitespace=True)

    home: HttpUrl = Field(
        None,
        examples=["https://very-big-corp.com"],
    )
    mastodon: HttpUrl = Field(
        None,
        examples=["https://mastodon.social/@VeryBigCorpOfficial"],
    )
    github: HttpUrl = Field(
        None,
        examples=["https://github.com/Very-Big-Corp"],
    )
    linkedin: HttpUrl = Field(
        None,
        examples=["https://linkedin.com/company/very-big-corp"],
    )
    twitter: HttpUrl = Field(
        None,
        examples=["https://twitter.com/VeryBigCorp"],
    )
    facebook: HttpUrl = Field(
        None,
        examples=["https://facebook.com/Very-Big-Corp"],
    )
