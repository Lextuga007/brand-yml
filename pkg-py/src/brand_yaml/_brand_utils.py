from __future__ import annotations

import logging
from copy import deepcopy

from pydantic import BaseModel, ConfigDict, Field
from typing import Any, TypeVar, Generic, Optional

T = TypeVar("T")


class BrandLightDark(BaseModel, Generic[T]):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    light: T = None
    dark: T = None


class BrandLightDarkString(BrandLightDark[str]):
    pass


class BrandWith(BaseModel, Generic[T]):
    model_config = ConfigDict(
        extra="ignore",
        str_strip_whitespace=True,
        populate_by_name=True,
        revalidate_instances="always",
    )

    with_: Optional[dict[str, T]] = Field(default=None, alias="with")

    def model_post_init(self, __context: Any) -> None:
        if self.with_ is None:
            return

        logging.debug("resolving with_ values")
        self._replace_with_recursively()

    def __setattr__(self, name: str, value: Any) -> None:
        super().__setattr__(name, value)
        self._replace_with_recursively()

    def _replace_with_recursively(self, items: dict | BaseModel | None = None, level=0):
        if level > 50:
            logging.error("BrandWith recursion limit reached")
            return

        if items is None:
            items = self

        if not isinstance(items, (dict, BaseModel)):
            return

        if isinstance(items, BaseModel):
            items_keys = items.model_fields.keys()
        elif hasattr(items, "keys"):
            items_keys = items.keys()

        for key in items_keys:
            logging.debug(f"checking key {key}")

            if isinstance(items, BaseModel):
                value = getattr(items, key)
            elif isinstance(items, dict):
                value = items[key]

            if isinstance(value, str) and value in self.with_:
                logging.debug(f"replacing key {key}")
                if isinstance(items, BaseModel):
                    setattr(items, key, deepcopy(self.with_[value]))
                elif isinstance(items, dict):
                    items[key] = deepcopy(self.with_[value])
            elif isinstance(value, (dict, BaseModel)):
                # TODO: we may want to avoid recursing into child BrandWith instances
                logging.debug(f"recursing into {key}")
                self._replace_with_recursively(value, level + 1)
            else:
                logging.debug(
                    f"skipping {key}, not replaceable (or not a dict or pydantic model)"
                )
