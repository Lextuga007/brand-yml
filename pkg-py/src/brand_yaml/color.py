from __future__ import annotations

from copy import deepcopy
from typing import Optional

from pydantic import (
    ConfigDict,
    field_validator,
    model_validator,
)

from ._defs import check_circular_references, defs_replace_recursively
from .base import BrandBase


class BrandColor(BrandBase):
    """
    Brand Colors

    The brand's custom color palette and theme.
    """

    model_config = ConfigDict(
        extra="forbid",
        revalidate_instances="always",
        validate_assignment=True,
        use_attribute_docstrings=True,
    )

    palette: dict[str, str] | None = None

    foreground: Optional[str] = None
    """The foreground color, used for text."""

    background: Optional[str] = None
    """The background color, used for the page or main background."""

    primary: Optional[str] = None
    """
    The primary accent color, i.e. the main theme color. Typically used for
    hyperlinks, active states, primary action buttons, etc.
    """

    secondary: Optional[str] = None
    """
    The secondary accent color. Typically used for lighter text or disabled
    states.
    """

    tertiary: Optional[str] = None
    """
    The tertiary accent color. Typically an even lighter color, used for
    hover states, accents, and wells.
    """

    success: Optional[str] = None
    """The color used for positive or successful actions and information."""

    info: Optional[str] = None
    """The color used for neutral or informational actions and information."""

    warning: Optional[str] = None
    """The color used for warning or cautionary actions and information."""

    danger: Optional[str] = None
    """The color used for errors, dangerous actions, or negative information."""

    light: Optional[str] = None
    """
    A bright color, used as a high-contrast foreground color on dark elements
    or low-contrast background color on light elements.
    """

    dark: Optional[str] = None
    """
    A dark color, used as a high-contrast foreground color on light elements
    or high-contrast background color on light elements.
    """

    emphasis: Optional[str] = None
    """A color used to emphasize or highlight text or elements."""

    link: Optional[str] = None
    """
    The color used for hyperlinks. If not defined, the `primary` color is
    used.
    """

    @field_validator("palette")
    @classmethod
    def create_brand_palette(cls, value: dict[str, str] | None):
        if value is None:
            return

        if not isinstance(value, dict):
            raise ValueError("`palette` must be a dictionary")

        check_circular_references(value)
        # We resolve `color.palette` on load or on replacement only
        # TODO: Replace with class with getter/setters
        #       Retain original values, return resolved values, and re-validate on update.
        defs_replace_recursively(value, name="palette")

        return value

    def _color_defs(self, resolved: bool = False) -> dict[str, str]:
        defs = deepcopy(self.palette) if self.palette is not None else {}
        defs.update(
            {
                k: v
                for k, v in self.model_dump().items()
                if k != "palette" and v is not None
            }
        )

        if resolved:
            defs_replace_recursively(defs)
            return defs
        else:
            return defs

    @model_validator(mode="after")
    def resolve_palette_values(self):
        defs_replace_recursively(
            self,
            defs=self._color_defs(resolved=False),
            name="color",
            exclude="palette",
        )
        return self
