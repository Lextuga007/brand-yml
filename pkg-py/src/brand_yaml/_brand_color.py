from __future__ import annotations

from typing import Optional

from pydantic import ConfigDict, Field

from ._brand_utils import BrandWith


class BrandColor(BrandWith[str]):
    """
    Brand Colors

    The brand's custom color palette and theme.

    Attributes
    ----------

    foreground
        The foreground color, used for text.
    
    background
        The background color, used for the page or main background.
    
    primary
        The primary accent color, i.e. the main theme color. Typically used for
        hyperlinks, active states, primary action buttons, etc.
    
    secondary
        The secondary accent color. Typically used for lighter text or disabled
        states.
    
    tertiary
        The tertiary accent color. Typically an even lighter color, used for
        hover states, accents, and wells.
    
    success
        The color used for positive or successful actions and information.
    
    info
        The color used for neutral or informational actions and information.
    
    warning
        The color used for warning or cautionary actions and information.
    
    danger
        The color used for errors, dangerous actions, or negative information.
    
    light
        A bright color, used as a high-contrast foreground color on dark
        elements or low-contrast background color on light elements.
    
    dark
        A dark color, used as a high-contrast foreground color on light elements
        or high-contrast background color on light elements.
    
    emphasis
        A color used to emphasize or highlight text or elements.
    
    link
        The color used for hyperlinks. If not defined, the `primary` color is
        used.

    """
    model_config = ConfigDict(extra="forbid")

    foreground: Optional[str] = Field(
        default=None,
        description="The foreground color, used for text.",
    )
    background: Optional[str] = Field(
        default=None,
        description="The background color, used for the page or main background.",
    )
    primary: Optional[str] = Field(
        default=None,
        description="The primary accent color, i.e. the main theme color. Typically used for hyperlinks, active states, primary action buttons, etc.",
    )
    secondary: Optional[str] = Field(
        default=None,
        description="The secondary accent color. Typically used for lighter text or disabled states.",
    )
    tertiary: Optional[str] = Field(
        default=None,
        description="The tertiary accent color. Typically an even lighter color, used for hover states, accents, and wells.",
    )

    success: Optional[str] = Field(
        default=None,
        description="The color used for positive or successful actions and information.",
    )
    info: Optional[str] = Field(
        default=None,
        description="The color used for neutral or informational actions and information.",
    )
    warning: Optional[str] = Field(
        default=None,
        description="The color used for warning or cautionary actions and information.",
    )
    danger: Optional[str] = Field(
        default=None,
        description="The color used for errors, dangerous actions, or negative information.",
    )

    light: Optional[str] = Field(
        default=None,
        description="A bright color, used as a high-contrast foreground color on dark elements or low-contrast background color on light elements.",
    )
    dark: Optional[str] = Field(
        default=None,
        description="A dark color, used as a high-contrast foreground color on light elements or high-contrast background color on light elements.",
    )
    emphasis: Optional[str] = Field(
        default=None,
        description="A color used to emphasize or highlight text or elements.",
    )
    link: Optional[str] = Field(
        default=None,
        description="The color used for hyperlinks. If not defined, the `primary` color is used.",
    )
