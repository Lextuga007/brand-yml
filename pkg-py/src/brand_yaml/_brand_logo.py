from __future__ import annotations

from typing import Union

from pydantic import ConfigDict
from ._brand_utils import BrandStringLightDark, BrandWith

class BrandLogo(BrandWith[Union[str, BrandStringLightDark]]):
    model_config = ConfigDict(extra="forbid")

    small: str | BrandStringLightDark = None
    medium: str | BrandStringLightDark = None
    large: str | BrandStringLightDark = None

