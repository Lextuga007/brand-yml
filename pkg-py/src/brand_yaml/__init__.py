from __future__ import annotations

from pathlib import Path
from typing import Any, Literal, overload

from pydantic import BaseModel, ConfigDict, Field, model_validator
from ruamel.yaml import YAML

from ._path import FileLocation
from ._utils import find_project_brand_yaml, recurse_dicts_and_models
from .color import BrandColor
from .logo import BrandLogo
from .meta import BrandMeta
from .typography import BrandTypography

yaml = YAML()


class Brand(BaseModel):
    model_config = ConfigDict(
        extra="ignore",
        revalidate_instances="always",
        validate_assignment=True,
    )

    meta: BrandMeta | None = Field(None)
    logo: str | BrandLogo | None = Field(None)
    color: BrandColor | None = Field(None)
    typography: BrandTypography | None = Field(None)
    defaults: dict[str, Any] | None = Field(None)
    path: Path | None = Field(None, exclude=True, repr=False)

    @classmethod
    def from_yaml(cls, path: str | Path):
        """
        Read a brand YAML file

        Reads a brand YAML file or finds and reads a `_brand.yml` file and returns
        a validated :class:`Brand` object.

        Parameters
        ----------
        path
            The path to the brand YAML file or a directory where `_brand.yml` is
            expected to be found. Typically, you can pass `__file__` from the
            calling script to find `_brand.yml` in the current directory or any of
            its parent directories.

        Returns
        -------
        :
            A validated :class:`Brand` object with all fields populated according to
            the brand YAML file.

        Raises
        ------
        :
            Raises a `FileNotFoundError` if no brand configuration file is found
            within the given path. Raises `ValueError` or other validation errors
            from [pydantic](https://docs.pydantic.dev/latest/) if the brand YAML
            file is invalid.

        Examples
        --------

        ```python
        from brand_yaml import Brand

        brand = Brand.from_yaml(__file__)
        brand = Brand.from_yaml("path/to/_brand.yml")
        ```
        """
        return cls.model_validate(read_brand_yaml(path, as_data=True))

    @model_validator(mode="after")
    def resolve_typography_colors(self):
        if self.typography is None or self.color is None:
            return self

        color_defs = self.color._color_defs(resolved=True)

        for top_field in self.typography.model_fields.keys():
            typography_node = getattr(self.typography, top_field)

            if not isinstance(typography_node, BaseModel):
                continue

            for typography_node_field in typography_node.model_fields.keys():
                if typography_node_field not in ("color", "background_color"):
                    continue

                value = getattr(typography_node, typography_node_field)
                if value is None or not isinstance(value, str):
                    continue

                if value not in color_defs:
                    continue

                setattr(
                    typography_node,
                    typography_node_field,
                    color_defs[value],
                )

        return self

    @model_validator(mode="after")
    def resolve_paths(self):
        path = self.path
        if path is not None:
            recurse_dicts_and_models(
                self,
                pred=lambda value: isinstance(value, FileLocation),
                modify=lambda value: value.set_root_dir(
                    path.parent,
                    validate_path=True,
                ),
            )
        return self


@overload
def read_brand_yaml(
    path: str | Path, as_data: Literal[False] = False
) -> Brand: ...


@overload
def read_brand_yaml(path: str | Path, as_data: Literal[True]) -> dict: ...


def read_brand_yaml(path: str | Path, as_data: bool = False) -> Brand | dict:
    """
    Read a brand YAML file

    Reads a brand YAML file or finds and reads a `_brand.yml` file and returns
    a validated :class:`Brand` object.

    Parameters
    ----------
    path
        The path to the brand YAML file or a directory where `_brand.yml` is
        expected to be found. Typically, you can pass `__file__` from the
        calling script to find `_brand.yml` in the current directory or any of
        its parent directories.

    as_data
        When `True`, returns the raw brand data as a dictionary parsed from the
        YAML file. When `False`, returns a validated :class:`Brand` object.

    Returns
    -------
    :
        A validated :class:`Brand` object with all fields populated according to
        the brand YAML file (`as_data=False`, default) or the raw brand data
        as a dictionary (`as_data=True`).

    Raises
    ------
    :
        Raises a `FileNotFoundError` if no brand configuration file is found
        within the given path. Raises `ValueError` or other validation errors
        from [pydantic](https://docs.pydantic.dev/latest/) if the brand YAML
        file is invalid.

    Examples
    --------

    ```python
    from brand_yaml import read_brand_yaml

    brand = read_brand_yaml(__file__)
    brand = read_brand_yaml("path/to/_brand.yml")
    ```
    """

    path = Path(path).absolute()

    if path.is_dir():
        path = find_project_brand_yaml(path)
    elif path.suffix == ".py":
        # allows users to simply pass `__file__`
        path = find_project_brand_yaml(path.parent)

    with open(path, "r") as f:
        brand_data = yaml.load(f)

    if not isinstance(brand_data, dict):
        raise ValueError(
            f"Invalid brand YAML file {str(path)!r}. Must be a dictionary."
        )

    brand_data["path"] = path

    if as_data:
        return brand_data

    return Brand.model_validate(brand_data)


__all__ = [
    "Brand",
    "read_brand_yaml",
]
