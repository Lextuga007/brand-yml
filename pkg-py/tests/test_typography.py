from __future__ import annotations

from urllib.parse import unquote

import pytest
from brand_yaml import read_brand_yaml
from brand_yaml.color import BrandColor
from brand_yaml.typography import (
    BrandTypography,
    BrandTypographyBase,
    BrandTypographyFontBunny,
    BrandTypographyFontFiles,
    BrandTypographyFontFilesPath,
    BrandTypographyFontGoogle,
    BrandTypographyGoogleFontsApi,
    BrandTypographyHeadings,
    BrandTypographyLink,
    BrandTypographyMonospace,
    BrandTypographyMonospaceBlock,
    BrandTypographyMonospaceInline,
)
from syrupy.extensions.json import JSONSnapshotExtension
from utils import path_examples, pydantic_data_from_json


@pytest.fixture
def snapshot_json(snapshot):
    return snapshot.use_extension(JSONSnapshotExtension)


@pytest.mark.parametrize(
    "path, fmt",
    [
        ("my-font.otf", "opentype"),
        ("my-font.ttf", "truetype"),
        ("my-font.woff", "woff"),
        ("my-font.woff2", "woff2"),
    ],
)
def test_brand_typography_font_file_format(path, fmt):
    font = BrandTypographyFontFilesPath(path=path)

    assert font.path == path
    assert font.format == fmt


def test_brand_typography_font_file_weight():
    args = {
        "path": "my-font.otf",
    }

    with pytest.raises(ValueError):
        BrandTypographyFontFilesPath.model_validate(
            {**args, "weight": "invalid"}
        )

    with pytest.raises(ValueError):
        BrandTypographyFontFilesPath.model_validate({**args, "weight": 999})

    with pytest.raises(ValueError):
        BrandTypographyFontFilesPath.model_validate({**args, "weight": 150})

    with pytest.raises(ValueError):
        BrandTypographyFontFilesPath.model_validate({**args, "weight": 0})

    assert (
        BrandTypographyFontFilesPath.model_validate(
            {**args, "weight": 100}
        ).weight
        == 100
    )
    assert (
        BrandTypographyFontFilesPath.model_validate(
            {**args, "weight": "thin"}
        ).weight
        == 100
    )
    assert (
        BrandTypographyFontFilesPath.model_validate(
            {**args, "weight": "semi-bold"}
        ).weight
        == 600
    )
    assert (
        BrandTypographyFontFilesPath.model_validate(
            {**args, "weight": "bold"}
        ).weight
        == "bold"
    )
    assert (
        BrandTypographyFontFilesPath.model_validate(
            {**args, "weight": "normal"}
        ).weight
        == "normal"
    )


def test_brand_typography_monospace():
    bt = BrandTypography.model_validate(
        {
            "monospace": {"family": "Fira Code", "size": "1.2rem"},
            "monospace-inline": {"size": "0.9rem"},
            "monospace-block": {
                "family": "Menlo",
            },
        }
    )

    assert bt.monospace is not None
    assert bt.monospace.family == "Fira Code"
    assert bt.monospace.size == "1.2rem"

    assert bt.monospace_inline is not None
    assert bt.monospace_inline.family == "Fira Code"  # inherits family
    assert bt.monospace_inline.size == "0.9rem"  # overrides size

    assert bt.monospace_block is not None
    assert bt.monospace_block.family == "Menlo"  # overrides family
    assert bt.monospace_block.size == "1.2rem"  # inherits size


def test_brand_typography_fields_base():
    base_fields = set(BrandTypographyBase.model_fields.keys())

    # TODO: Compare directly with brand-yaml spec
    assert base_fields == {
        "family",
        "weight",
        "size",
        "line_height",
        "color",
    }


def test_brand_typography_fields_headings():
    headings_fields = set(BrandTypographyHeadings.model_fields.keys())

    # TODO: Compare directly with brand-yaml spec
    assert headings_fields == {
        "family",
        "weight",
        "style",
        "line_height",
        "color",
    }


def test_brand_typography_fields_monospace():
    fields = set(BrandTypographyMonospace.model_fields.keys())

    # TODO: Compare directly with brand-yaml spec
    assert fields == {"family", "weight", "size"}


def test_brand_typography_fields_monospace_inline():
    fields = set(BrandTypographyMonospaceInline.model_fields.keys())

    # TODO: Compare directly with brand-yaml spec
    assert fields == {
        "family",
        "weight",
        "size",
        "color",
        "background_color",
    }


def test_brand_typography_fields_monospace_block():
    fields = set(BrandTypographyMonospaceBlock.model_fields.keys())

    # TODO: Compare directly with brand-yaml spec
    assert fields == {
        "family",
        "weight",
        "size",
        "line_height",
        "color",
        "background_color",
    }


def test_brand_typography_fields_link():
    fields = set(BrandTypographyLink.model_fields.keys())

    # TODO: Compare directly with brand-yaml spec
    assert fields == {
        "weight",
        "decoration",
        "color",
        "background_color",
    }


def test_brand_typography_font_bunny():
    bf = BrandTypography.model_validate(
        {
            "fonts": [
                {
                    "source": "bunny",
                    "family": "Kode Mono",
                    "weight": [400, 500, 600, 700],
                    "style": "normal",
                }
            ]
        }
    )

    assert len(bf.fonts) == 1
    assert isinstance(bf.fonts[0], BrandTypographyFontBunny)
    assert isinstance(bf.fonts[0], BrandTypographyGoogleFontsApi)


def test_brand_typography_font_google_import_url():
    bg = BrandTypography.model_validate(
        {
            "fonts": [
                {
                    "source": "google",
                    "family": "Open Sans",
                    "weight": [700, 400],
                    "style": ["italic", "normal"],
                }
            ]
        }
    )

    assert len(bg.fonts) == 1
    assert isinstance(bg.fonts[0], BrandTypographyFontGoogle)
    assert (
        unquote(bg.fonts[0].import_url())
        == "https://fonts.googleapis.com/css2?family=Open+Sans:ital,wght@0,400;0,700;1,400;1,700&display=auto"
    )


def test_brand_typography_font_bunny_import_url():
    bg = BrandTypography.model_validate(
        {
            "fonts": [
                {
                    "source": "bunny",
                    "family": "Open Sans",
                    "weight": [700, 400],
                    "style": ["italic", "normal"],
                }
            ]
        }
    )

    assert len(bg.fonts) == 1
    assert isinstance(bg.fonts[0], BrandTypographyFontBunny)
    assert (
        unquote(bg.fonts[0].import_url())
        == "https://fonts.bunny.net/css?family=Open+Sans:400,400i,700,700i&display=auto"
    )


def test_brand_typography_ex_simple(snapshot_json):
    brand = read_brand_yaml(path_examples("brand-typography-simple.yml"))

    assert isinstance(brand.typography, BrandTypography)

    assert isinstance(brand.typography.fonts, list)
    assert len(brand.typography.fonts) == 3
    assert [f.family for f in brand.typography.fonts] == [
        "Open Sans",
        "Roboto Slab",
        "Fira Code",
    ]
    assert [f.source for f in brand.typography.fonts] == ["google"] * 3

    assert brand.typography.link is None
    assert isinstance(brand.typography.base, BrandTypographyBase)
    assert isinstance(brand.typography.headings, BrandTypographyHeadings)
    assert isinstance(brand.typography.monospace, BrandTypographyMonospace)
    assert isinstance(
        brand.typography.monospace_inline, BrandTypographyMonospace
    )
    assert isinstance(
        brand.typography.monospace_block, BrandTypographyMonospace
    )
    assert isinstance(
        brand.typography.monospace_inline, BrandTypographyMonospaceInline
    )
    assert isinstance(
        brand.typography.monospace_block, BrandTypographyMonospaceBlock
    )

    assert snapshot_json == pydantic_data_from_json(brand)


def test_brand_typography_ex_fonts(snapshot_json):
    brand = read_brand_yaml(path_examples("brand-typography-fonts.yml"))

    assert isinstance(brand.typography, BrandTypography)
    assert len(brand.typography.fonts) == 4

    # Local Font Files
    local_font = brand.typography.fonts[0]
    assert isinstance(local_font, BrandTypographyFontFiles)
    assert local_font.source == "file"
    assert local_font.family == "Open Sans"
    for i, font in enumerate(local_font.files):
        assert isinstance(font, BrandTypographyFontFilesPath)
        assert str(font.path).startswith("Open-Sans")
        assert str(font.path).endswith(".ttf")
        assert font.format == "truetype"
        assert font.weight == ["bold", "normal"][i]
        assert font.style == ["normal", "italic"][i]

    # Online Font Files
    online_font = brand.typography.fonts[1]
    assert isinstance(online_font, BrandTypographyFontFiles)
    assert online_font.source == "file"
    assert online_font.family == "Closed Sans"
    for i, font in enumerate(online_font.files):
        assert isinstance(font, BrandTypographyFontFilesPath)
        assert str(font.path).startswith("https://")
        assert str(font.path).endswith(".woff2")
        assert font.format == "woff2"
        assert font.weight == ["bold", "normal"][i]
        assert font.style == ["normal", "italic"][i]

    # Google Fonts
    google_font = brand.typography.fonts[2]
    assert isinstance(google_font, BrandTypographyFontGoogle)
    assert google_font.family == "Roboto Slab"
    assert google_font.weight == 600
    assert google_font.style == "normal"
    assert google_font.display == "block"

    # Bunny Fonts
    bunny_font = brand.typography.fonts[3]
    assert isinstance(bunny_font, BrandTypographyFontBunny)
    assert bunny_font.family == "Fira Code"

    assert snapshot_json == pydantic_data_from_json(brand)


def test_brand_typography_ex_color(snapshot_json):
    brand = read_brand_yaml(path_examples("brand-typography-color.yml"))

    assert isinstance(brand.typography, BrandTypography)
    assert isinstance(brand.color, BrandColor)

    t = brand.typography
    color = brand.color
    assert color.palette is not None

    assert isinstance(t.base, BrandTypographyBase)
    assert t.base.color == color.foreground

    assert isinstance(t.headings, BrandTypographyHeadings)
    assert t.headings.color == color.primary

    assert isinstance(t.monospace_inline, BrandTypographyMonospaceInline)
    assert t.monospace_inline.color == color.background
    assert t.monospace_inline.background_color == color.palette["red"]

    assert isinstance(t.monospace_block, BrandTypographyMonospaceBlock)
    assert t.monospace_block.color == color.foreground
    assert t.monospace_block.background_color == color.background

    assert isinstance(t.link, BrandTypographyLink)
    assert t.link.color == color.palette["red"]

    assert snapshot_json == pydantic_data_from_json(brand)


def test_brand_typography_ex_minimal(snapshot_json):
    brand = read_brand_yaml(path_examples("brand-typography-minimal.yml"))

    assert isinstance(brand.typography, BrandTypography)

    assert isinstance(brand.typography.fonts, list)
    assert len(brand.typography.fonts) == 3
    assert brand.typography.fonts[0].source == "file"
    assert brand.typography.fonts[0].files == []

    assert isinstance(brand.typography.fonts[1], BrandTypographyFontGoogle)
    assert brand.typography.fonts[1].family == "Roboto Slab"

    assert isinstance(brand.typography.fonts[2], BrandTypographyFontGoogle)
    assert brand.typography.fonts[2].family == "Fira Code"

    assert snapshot_json == pydantic_data_from_json(brand)
