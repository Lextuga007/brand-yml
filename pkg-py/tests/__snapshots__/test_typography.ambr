# serializer version: 1
# name: test_brand_typography_css_fonts
  '''
  @font-face {
    font-family: 'Open Sans';
    font-weight: auto;
    font-style: normal;
    src: url('fonts/open-sans/OpenSans-Variable.ttf') format('truetype');
  }
  @font-face {
    font-family: 'Open Sans';
    font-weight: auto;
    font-style: italic;
    src: url('fonts/open-sans/OpenSans-Variable-Italic.ttf') format('truetype');
  }
  @font-face {
    font-family: 'Closed Sans';
    font-weight: bold;
    font-style: normal;
    src: url('https://example.com/Closed-Sans-Bold.woff2') format('woff2');
  }
  @font-face {
    font-family: 'Closed Sans';
    font-weight: auto;
    font-style: italic;
    src: url('https://example.com/Closed-Sans-Italic.woff2') format('woff2');
  }
  @import url('https://fonts.googleapis.com/css2?family=Roboto+Slab%3Aital%2Cwght%400%2C600&display=block');
  @import url('https://fonts.bunny.net/css?family=Fira+Code%3A100%2C100i%2C200%2C200i%2C300%2C300i%2C400%2C400i%2C500%2C500i%2C600%2C600i%2C700%2C700i%2C800%2C800i%2C900%2C900i&display=auto');
  '''
# ---
